### DISCUSSION TITLE
Verifying the integrity of a lockfile

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3202] [Ideas]
We cannot guarantee that Git or other version control system merges changes to lockfiles correctly. Issues may happen sometimes (see https://github.com/pnpm/pnpm/issues/3188, https://github.com/pnpm/pnpm/issues/3137). Hence, we need pnpm to verify the correctness of the data which is present in the lockfile.

Currently, pnpm blindly trusts any lockfile that is up-to-date with the `package.json` file of the repository. This allows pnpm to be extremely fast. However, this also means that in case of an invalid lockfile, pnpm will create an invalid modules directory.

Possible solution: we could add a new field to the lockfile called `lockfileChecksum`. This field will contain a hash of the lockfile. When the hash will be out-of-date, pnpm will not run the fast install algorithm but will instead revalidate the data in the lockfile.

Cons of this solution. After any merge/rebase that modifies the lockfile, `pnpm install` will have to be executed to revalidate the lockfile.

cc @pnpm/collaborators

**Comment by statianzo** [comment]
> Currently, pnpm blindly trusts any lockfile that is up-to-date with the package.json file of the repository.

Meaning it do a shallow comparison of top-level modules?

---

Rather than rebuilding the whole lockfile, does something like a [Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) make sense here? Could pnpm detect a branch that's out of date without discarding everything?

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbnJvb3RbXCJyb290LXBhY2thZ2VAMC4wLjFcIl1cbnJvb3QgLS0-IGExW1wiQUAxLjBcIl1cbnJvb3QgLS0-IGIyW1wiQkAyLjBcIl1cbnJvb3QgLS0-IGMzW1wiQ0AzLjBcIl1cbmIyIC0tPiBjMltcIkNAMi4wXCJdXG5hMSAtLT4gYzJbXCJDQDIuMFwiXVxuYzIgLS0-IGQzW1wiREAzLjBcIl1cbmMyIC0tPiBlNFtcIkVANC4wXCJdXG5jMyAtLT4gZTRcbmMzIC0tPiBmMVtcIkZAMS4wICh0YW1wZXJlZClcIl1cblxuc3R5bGUgZjEgZmlsbDogI0ZGQUE5OVxuc3R5bGUgYzMgZmlsbDogI0ZGQUE5OVxuc3R5bGUgcm9vdCBmaWxsOiAjRkZBQTk5IiwibWVybWFpZCI6e30sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbnJvb3RbXCJyb290LXBhY2thZ2VAMC4wLjFcIl1cbnJvb3QgLS0-IGExW1wiQUAxLjBcIl1cbnJvb3QgLS0-IGIyW1wiQkAyLjBcIl1cbnJvb3QgLS0-IGMzW1wiQ0AzLjBcIl1cbmIyIC0tPiBjMltcIkNAMi4wXCJdXG5hMSAtLT4gYzJbXCJDQDIuMFwiXVxuYzIgLS0-IGQzW1wiREAzLjBcIl1cbmMyIC0tPiBlNFtcIkVANC4wXCJdXG5jMyAtLT4gZTRcbmMzIC0tPiBmMVtcIkZAMS4wICh0YW1wZXJlZClcIl1cblxuc3R5bGUgZjEgZmlsbDogI0ZGQUE5OVxuc3R5bGUgYzMgZmlsbDogI0ZGQUE5OVxuc3R5bGUgcm9vdCBmaWxsOiAjRkZBQTk5IiwibWVybWFpZCI6e30sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)

In the above, `root-package` is invalidated by the tampered F, but the `A@1.0` and `B@2.0` don't need to get rebuilt because their child hashes are still valid.


Whatever the implementation, having a means to verify this during CI and fail on checksum mismatch, would be useful. Otherwise, if a poor merge is made, every dev who pulls will be subject to pnpm in "slow mode".

**Reply by zkochan** [reply]
It would cause even bigger git diffs and the merges would result in conflicts more frequently.

**Reply by statianzo** [reply]
Good point. Stability of lock file was a major reason I adopted pnpm.

**Comment by zkochan** [comment]
Another solution might be to always check the actual name/version (or the package ID?) of the package inside the tarball with the name/version that is present in the lockfile. If they don't much, the lockfile is broken.

To implement this, we will probably need to add the name and version of the package to the package index file. The package index file is a file in the store that contains mappings of the package files to their locations in the content-addressable store ([for example](https://github.com/pnpm/sample-project/blob/4aedd25f8aa0dc3d7be458999f4b380ed8ec25a9/pnpm-store/v3/files/a8/115c55e4a702fe4d150abd3872822a7e09fc98-index.json)).

**Reply by statianzo** [reply]
Certainly a positive that a valid merge wouldn't force checksum recreation because integrity is not stored in the lockfile. What is the downside of this approach? Storage? Migrating from older versions of the structure?

> or the package ID?
That would help with git based dependencies, right?

**Comment by statianzo** [comment]
> Cons of this solution. After any merge/rebase that modifies the lockfile, pnpm install will have to be executed to revalidate the lockfile.

Having to `pnpm install` after any non-pnpm operations on the lockfile would be troublesome.

1. Dev A merges PR into main with pnpm-lock
2. Dev B merges PR into main with pnpm-lock, git merges fine
3. Devs C,D,E pull main and `pnpm i`
4. C, D, E all have a locally modified pnpm-lock with updated integrity

**Comment by vis97c** [comment]
Any update? I'm migrating from yarn which has "yarn check --integrity" that I use on my project as a precommit hook, quite useful when working with monorepos.

**Comment by cafesanu** [comment]
`pnpm install --frozen-lockfile` does the trick

**Reply by camsteffen** [reply]
For anyone who might see this, note `--frozen-lockfile` is enabled by default in CI.
